packages/core/src/main/conversation/ConversationService.ts

```diff
@@ -1045,126 +1077,104 @@ export class ConversationService {
-        genericMessage = this.generateTextGenericMessage(payloadBundle);
-        break;
-      default:
-        throw new Error(`No send method implemented for "${payloadBundle['type']}".`);
-    }
+    const coreCryptoClient = this.coreCryptoClientProvider();
+    const encrypted = await coreCryptoClient.encryptMessage(
+      groupIdBytes,
+      GenericMessage.encode(genericMessage).finish(),
+    );
 
-    if ((await callbacks?.onStart?.(genericMessage)) === false) {
-      // If the onStart call returns false, it means the consumer wants to cancel the message sending
-      return {...payloadBundle, state: PayloadBundleState.CANCELLED};
+    try {
+      await this.apiClient.api.conversation.postMlsMessage(encrypted);
+      onSuccess?.(genericMessage, new Date().toISOString());
+      return {
+        ...payload,
+        content,
+        messageTimer: genericMessage.ephemeral?.expireAfterMillis || 0,
+        state: PayloadBundleState.OUTGOING_SENT,
+      };
+    } catch (error) {
+      return {
+        ...payload,
+        content,
+        messageTimer: genericMessage.ephemeral?.expireAfterMillis || 0,
+        state: PayloadBundleState.CANCELLED,
+      };
     }
+  }
 
+  private async sendProteusMessage<T extends OtrMessage>(
+    params: SendProteusMessageParams<T>,
+    genericMessage: GenericMessage,
+    content: T['content'],
+  ): Promise<T> {
+    const {userIds, sendAsProtobuf, conversationDomain, nativePush, targetMode, payload, onClientMismatch, onSuccess} =
+      params;
     const response = await this.sendGenericMessage(
       this.apiClient.validatedClientId,
-      payloadBundle.conversation,
+      payload.conversation,
       genericMessage,
       {
         userIds,
         sendAsProtobuf,
         conversationDomain,
         nativePush,
         targetMode,
-        onClientMismatch: callbacks?.onClientMismatch,
+        onClientMismatch,
       },
     );
 
     if (!response.errored) {
       if (!this.isClearFromMismatch(response)) {
         // We warn the consumer that there is a mismatch that did not prevent message sending
-        await callbacks?.onClientMismatch?.(response, true);
+        await onClientMismatch?.(response, true);
       }
-      callbacks?.onSuccess?.(genericMessage, response.time);
+      onSuccess?.(genericMessage, response.time);
     }
 
     return {
-      ...payloadBundle,
-      content: processedContent || payloadBundle.content,
+      ...payload,
+      content,
       messageTimer: genericMessage.ephemeral?.expireAfterMillis || 0,
       state: response.errored ? PayloadBundleState.CANCELLED : PayloadBundleState.OUTGOING_SENT,
     };
   }
 
+  /**
+   * Sends a message to a conversation
+   *
+   * @param params.payload The message to send to the conversation
+   * @param params.protocol The protocol to use to send the message (MLS or Proteus)
+   * @param params.groupId? The groupId of the conversation to send the message to (Needed only for MLS)
```

Comment:
This doc can be moved to the SendMlsMessageParams. This will allow VScode to show the doc as you type ;)

Reply:
thanks https://github.com/wireapp/wire-web-packages/pull/4314/commits/1eca14a201ce41c0fcaab10131fe4cf3a593aa70